File: Source/AlphaTab.JavaScript/Utils/FontLoadingChecker.cs
using System;
using AlphaTab.Haxe.Js;
using AlphaTab.Haxe.Js.Html;
using AlphaTab.Platform;
using AlphaTab.Util;
using Phase;

namespace AlphaTab.Utils
{
    /// <summary>
    /// This small utility helps to detect whether a particular font is already loaded.
    /// </summary>
    internal class FontLoadingChecker
    {
        private readonly string _family;
        private readonly string _fallbackText;
        private bool _isStarted;
        public bool IsFontLoaded { get; set; }
        public event Action FontLoaded;

        private void OnFontLoaded()
        {
            var handler = FontLoaded;
            if (handler != null)
            {
                handler();
            }
        }

        public FontLoadingChecker(
            string family,
            string fallbackText = null)
        {
            _family = family;
            _fallbackText = fallbackText == null ? "BESbwy" : fallbackText;
        }

        public void CheckForFontAvailability()
        {
            var isWorker =
                Script.Write<bool>(
                    "untyped __js__(\"typeof(WorkerGlobalScope) !== 'undefined' && self instanceof WorkerGlobalScope\")");
            if (isWorker)
            {
                // no web fonts in web worker
                IsFontLoaded = false;
                return;
            }

            if (_isStarted)
            {
                return;
            }

            _isStarted = true;
            var failCounter = 0;
            var failCounterId = Browser.Window.SetInterval(new Action(() =>
            {
                failCounter++;
                Logger.Warning("Rendering",
                    "Could not load font '" + _family + "' within " + (failCounter * 5) + " seconds");
            }),
            5000);

            Logger.Debug("Font", "Start checking for font availablility: " + _family);
            var cssFontLoadingModuleSupported = Browser.Document.Fonts.IsTruthy() &&
                                                Browser.Document.Fonts.Member<object>("load").IsTruthy();
            if (cssFontLoadingModuleSupported)
            {
                Logger.Debug("Font", "[" + _family + "] Font API available");
                Action checkFont = null;
                checkFont = () =>
                {
                    Browser.Document.Fonts.Load("1em " + _family).Then(_ =>
                    {
                        Logger.Debug("Font", "[" + _family + "] Font API signaled loaded");
                        if (Browser.Document.Fonts.Check("1em " + _family))
                        {
                            Logger.Info("Rendering", "[" + _family + "] Font API signaled available");
                            IsFontLoaded = true;
                            Browser.Window.ClearInterval(failCounterId);
                            OnFontLoaded();
                        }
                        else
                        {
                            Logger.Debug("Font",
                                "[" + _family +
                                "] Font API loaded reported, but font not available, checking later again");
                            Browser.Window.SetTimeout((Action)(() =>
                                {
                                    checkFont();
                                }),
                                250);
                        }

                        return true;
                    });
                };
                checkFont();
            }
            else
            {
                Logger.Debug("Font", "[" + _family + "] Font API not available, using resize trick");
                // based on the idea of https://www.bramstein.com/writing/detecting-system-fonts-without-flash.html
                // simply create 3 elements with the 3 default font families and measure them
                // then change to the desired font and expect a change on the width

                Element sans = null;
                Element serif = null;
                Element monospace = null;

                var initialSansWidth = -1;
                var initialSerifWidth = -1;
                var initialMonospaceWidth = -1;

                Action checkFont = null;
                checkFont = () =>
                {
                    if (sans == null)
                    {
                        sans = CreateCheckerElement("sans-serif");
                        serif = CreateCheckerElement("serif");
                        monospace = CreateCheckerElement("monospace");

                        Browser.Document.Body.AppendChild(sans);
                        Browser.Document.Body.AppendChild(serif);
                        Browser.Document.Body.AppendChild(monospace);

                        initialSansWidth = sans.OffsetWidth;
                        initialSerifWidth = serif.OffsetWidth;
                        initialMonospaceWidth = monospace.OffsetWidth;

                        sans.Style.FontFamily = "'" + _family + "',sans-serif";
                        serif.Style.FontFamily = "'" + _family + "',serif";
                        monospace.Style.FontFamily = "'" + _family + "',monospace";


                    }

                    var sansWidth = sans.OffsetWidth;
                    var serifWidth = serif.OffsetWidth;
                    var monospaceWidth = monospace.OffsetWidth;

                    if ((sansWidth != initialSansWidth && serifWidth != initialSerifWidth) ||
                        (sansWidth != initialSansWidth && monospaceWidth != initialMonospaceWidth) ||
                        (serifWidth != initialSerifWidth && monospaceWidth != initialMonospaceWidth))
                    {
                        if (sansWidth == serifWidth || sansWidth == monospaceWidth || serifWidth == monospaceWidth)
                        {
                            Browser.Document.Body.RemoveChild(sans);
                            Browser.Document.Body.RemoveChild(serif);
                            Browser.Document.Body.RemoveChild(monospace);
                            IsFontLoaded = true;
                            Browser.Window.ClearInterval(failCounterId);
                            OnFontLoaded();
                        }
                        else
                        {
                            Browser.Window.SetTimeout(checkFont, 250);
                        }
                    }
                    else
                    {
                        Browser.Window.SetTimeout(checkFont, 250);
                    }
                };

                Browser.Window.AddEventListener("DOMContentLoaded",
                    (Action)(() =>
                    {
                        checkFont();
                    }));
            }
        }

        private Element CreateCheckerElement(string family)
        {
            var document = Browser.Document;
            var checkerElement = document.CreateElement("span");
            checkerElement.Style.Display = "inline-block";
            checkerElement.Style.Position = "absolute";
            checkerElement.Style.Overflow = "hidden";

            checkerElement.Style.Top = "-1000px";
            checkerElement.Style.FontSynthesis = "none";

            checkerElement.Style.FontSize = "100px";
            checkerElement.Style.FontFamily = family;
            checkerElement.InnerHTML = _fallbackText;

            document.Body.AppendChild(checkerElement);
            return checkerElement;
        }
    }
}

File: src/util/FontLoadingChecker.ts
import { IEventEmitterOfT, EventEmitterOfT } from '@src/EventEmitter';
import { Logger } from '@src/Logger';
import { Environment } from '@src/Environment';

/**
 * This small utility helps to detect whether a particular font is already loaded.
 * @target web
 */
export class FontLoadingChecker {
    private _originalFamilies: string[];
    private _families: string[];

    private _isStarted: boolean = false;
    public isFontLoaded: boolean = false;

    public fontLoaded: IEventEmitterOfT<string> = new EventEmitterOfT<string>();

    public constructor(families: string[]) {
        this._originalFamilies = families;
        this._families = families;
    }

    public checkForFontAvailability(): void {
        if (Environment.isRunningInWorker) {
            // no web fonts in web worker
            this.isFontLoaded = false;
            return;
        }

        if (this._isStarted) {
            return;
        }

        this._isStarted = true;
        let failCounter: number = 0;
        let failCounterId: number = window.setInterval(() => {
            Logger.warning(
                'Rendering',
                `Could not load font '${this._families[0]}' within ${(failCounter + 1) * 5} seconds`,
                null
            );

            // try loading next font if there are more than 1 left
            if (this._families.length > 1) {
                this._families.shift();
                failCounter = 0;
            } else {
                failCounter++;
            }
        }, 5000);

        Logger.debug('Font', `Start checking for font availablility: ${this._families.join(', ')}`);

        let errorHandler = (e: unknown) => {
            if (this._families.length > 1) {
                Logger.debug('Font', `[${this._families[0]}] Loading Failed, switching to ${this._families[1]}`, e);
                this._families.shift();
                window.setTimeout(() => {
                    // tslint:disable-next-line: no-floating-promises
                    checkFont();
                }, 0);
            } else {
                Logger.error('Font', `[${this._originalFamilies.join(',')}] Loading Failed, rendering cannot start`, e);
                window.clearInterval(failCounterId);
                debugger;
            }
        };

        let successHandler = (font: string) => {
            Logger.debug('Font', `[${font}] Font API signaled available`);
            this.isFontLoaded = true;
            window.clearInterval(failCounterId);
            (this.fontLoaded as EventEmitterOfT<string>).trigger(this._families[0]);
        };

        let checkFont = async () => {
            // Fast Path: check if one of the specified fonts is already available.
            for (const font of this._families) {
                if (await this.isFontAvailable(font, false)) {
                    successHandler(font);
                    return;
                }
            }

            // Slow path: Wait for fonts to be loaded sequentially
            try {
                await (document as any).fonts.load(`1em ${this._families[0]}`);
            } catch (e) {
                errorHandler(e);
            }

            Logger.debug('Font', `[${this._families[0]}] Font API signaled loaded`);
            if (await this.isFontAvailable(this._families[0], true)) {
                successHandler(this._families[0]);
            } else {
                errorHandler('Font not available');
            }
            return true;
        };

        (document as any).fonts.ready.then(() => {
            // tslint:disable-next-line: no-floating-promises
            checkFont();
        });
    }

    private isFontAvailable(family: string, advancedCheck: boolean): Promise<boolean> {
        return new Promise<boolean>(resolve => {
            // In some very rare occasions Chrome reports false for the font.
            // in this case we try to force some refresh and reload by creating an element with this font.
            const fontString = '1em ' + family;
            if ((document as any).fonts.check(fontString)) {
                resolve(true);
            } else if (advancedCheck) {
                Logger.debug('Font', `Font ${family} not available, creating test element to trigger load`);
                const testElement = document.createElement('div');
                testElement.style.font = fontString;
                testElement.style.opacity = '0';
                testElement.style.position = 'absolute';
                testElement.style.top = '0';
                testElement.style.left = '0';
                testElement.innerText = `Trigger ${family} load`;
                document.body.appendChild(testElement);
                setTimeout(() => {
                    document.body.removeChild(testElement);
                    if ((document as any).fonts.check(fontString)) {
                        resolve(true);
                    } else {
                        resolve(false);
                    }
                }, 200);
            } else {
                resolve(false);
            }
        });
    }
}
